### PR TITLE
Fix deepspeed prefix

### PR DESCRIPTION
**Pull request by thomasw21**
deepspeed implementation would prevent one from passing down an attention mask. We need this feature for prefix lm which has an attention mask that's batch dependent.

### CODE REVIEW

File: megatron/model/gpt_model.py [hunk 1]
```diff
@@ -210,6 +218,10 @@ def _to_float16(inputs):
                     self_attn_mask_type=AttnMaskType.prefix if prefix_lm else AttnMaskType.causal))
                 
         
+        if not hasattr(args, 'attn_mask'):
+            # We drop attention mask from the pipeline
+            self.specs.append(lambda x: x[0])
```
Comment (thomasw21) [inline megatron/model/gpt_model.py:223]: Unfortunately we drop attention here ....

File: megatron/model/language_model.py [hunk 2]
```diff
@@ -290,7 +290,6 @@ def forward(self, inputs, **kwargs):
         if hasattr(self._args, 'attn_mask'):
             return embeddings
         else:
-            assert False
```
Comment (thomasw21) [inline megatron/model/language_model.py:293]: We remove this in order to allow this case.

File: pretrain_gpt.py [hunk 3]
```diff
@@ -73,6 +65,13 @@ def model_provider(pre_process=True, post_process=True):
             # must be bool or the training crashes expecting bool, but getting Half
             args.attn_mask = attention_mask.to(torch.bool)
 
+            model = GPTModelPipe(
+                num_tokentypes=0,
+                parallel_output=True
+            )
+            # This is a hack to give us a reference to get_batch_pipe from within training.py
+            # We need to call model.set_batch_fn after deepspeed.initialize
+            model._megatron_batch_fn = get_batch_pipe
```
Comment (thomasw21) [inline pretrain_gpt.py:74]: We move this part of the code after setting attention mask in `args`. This allows us to use args while building the model, typically to distinguish gpt vs prefix-lm.

**Comment by stas00** [comment]
@thomasw21, is there a way to fix it directly in deepspeed? or do you feel this is good enough?

**Comment by thomasw21** [comment]
I think we're going to have to fix something in deepspeed somewhere. I'm still unclear on the errors I'm getting ...

https://huggingface.slack.com/archives/C01NHER1JLS/p1631861191377400

File: pretrain_prefix_lm.py [hunk 4]
```diff
@@ -33,6 +33,9 @@
 import os
 import subprocess
 
+class GPT2ModelPipe(GPTModelPipe):
```
[the remark this answers is not in the document]
Comment (stas00) [reply pretrain_prefix_lm.py:36]: If you could explain - how does your hack work, @thomasw21? 

--------------

Yes, this is one of the very limiting features of most pipe implementations I have seen - not just in Deepspeed - the new pytorch API is much more flexible wrt to the flexibility of passing the different types of inputs.

When I was trying to implement PP support in transformers I created all kinds of encoding/decoding wrappers to cope with this.  It's not pretty.

>I think the original reason why there's this issue, is torch.distributed doesn't handle bool tensors very well, but I think it's fixed in pytorch/pytorch#24137 . Maybe since our setup has a higher version, maybe we can remove all code linked to bool tensors?

I agree - this should have been fixed in pt-1.7 according to the link you shared.

Though the problem is that DS has to support older version, so it's tricky.

@ShadenSmith - do you have any good suggestions?

[the remark this answers is not in the document]
Comment (stas00) [reply pretrain_prefix_lm.py:36]: Ah, I see, I missed that it checks for a specific class name to activate different functionality. Thank you for explaining, Thomas.

I hope @ShadenSmith will have a moment to recommend a solution here, since he is the author of that code I think.

[the remark this answers is not in the document]
Comment (ShadenSmith) [reply pretrain_prefix_lm.py:36]: Thanks for the ping, and sorry for the trouble with these hacks.

I'm inclined to remove the hacks and just expect folks to either use torch 1.7+, or convert their activations from unsupported types at the model level.

We can support much more flexible message shapes/types with a more robust meta-communication. There some [p2p routines](https://github.com/microsoft/DeepSpeed/blob/big-science/deepspeed/runtime/pipe/p2p.py#L98) that serialize the message with `pickle()` and then just send that over to exchange the meta-information. Using `pickle` has a few tradeoffs, including an expensive GPU -> CPU -> GPU transfer for the first activation exchange, and the typical security risks associated with pickling.

[the remark this answers is not in the document]
Comment (stas00) [reply pretrain_prefix_lm.py:36]: but in this particular case it's just a tensor so it should be easy to just allow it if we ask for pt>=1.7

> We can support much more flexible message shapes/types with a more robust meta-communication. There some p2p routines that serialize the message with pickle() and then just send that over to exchange the meta-information. Using pickle has a few tradeoffs, including an expensive GPU -> CPU -> GPU transfer for the first activation exchange, and the typical security risks associated with pickling.

That would be great! but we can think of advanced features separately - as I mentioned in the past when porting PP - this would be extremely helpful to anybody wanting to integrated DS's PP into their software.

So just allowing bool tensors is the only showstopper at the moment if I understand it correctly. Thomas please correct me if I'm wrong.

[the remark this answers is not in the document]
Comment (thomasw21) [reply pretrain_prefix_lm.py:36]: Well I'm having a deadlock somewhere currently (or just sometging very slow). But at least no exceptions.

[the remark this answers is not in the document]
Comment (thomasw21) [reply pretrain_prefix_lm.py:36]: cc @StellaAthena @ontocord as you both expressed interest in deepspeed issues for prefix lm.

**Comment by thomasw21** [comment]
Small update:
 - If you install DeepSpeed using this [branch](https://github.com/thomasw21/DeepSpeed/tree/big-science-fix-passing-multiple-tensors) ([PR]()), you can run PP=1 for prefix lm. That for should still be stable for GPT models (at least I can run a GPT configuration).
 - PP>1 seems to deadlock somewhere around `optimiser.step` https://github.com/thomasw21/DeepSpeed/blob/big-science-fix-passing-multiple-tensors/deepspeed/runtime/engine.py#L1331 It's very unclear to me what is happening here:


Things I tried to fix PP>1:
 - switch optimizer to Adam instead of FusedAdam. This can be updated in this repo. No success, same symptoms.
 - tried to trigger `KeyboardInterrupt` in order to see where it deadlocks. I'm having issues viewing it as the launcher just not allows access to subprocess logs, ie it prints out "Killing subprocess {process_ID}".

File: tests/test_training.py [hunk 5]
```diff
@@ -174,3 +174,114 @@ def test_training_all(self):
         # test tensorboard (1 file from the first run, plus 1 now)
         tensorboard_files = glob.glob(f"{output_dir}/tensorboard/events*")
         self.assertEqual(len(tensorboard_files), 2, "tensorboard files")
+
+    @unittest.skip("Skip until deepspeed allows to pass tuple of tensors between pipeline stages.")
```
Comment (thomasw21) [inline tests/test_training.py:178]: Test passes with https://github.com/thomasw21/DeepSpeed/tree/big-science-fix-passing-multiple-tensors/deepspeed I'll wait to merge that PR on the big-science branch in order to un-skip that test.

**Comment by thomasw21** [comment]
So this PR is a partial fix of prefix lm, the other part is in `deepspeed` https://github.com/microsoft/DeepSpeed/pull/1400

I don't think this PR breaks anything, so I think it's ready for review!

File: tests/test_training.py (hunk 5 above)
Comment (stas00) [reply tests/test_training.py:178]: I did test with your branch and the test works.

Comment (stas00) [reply tests/test_training.py:178]: OK, I found it - you mentioned it elsewhere: https://github.com/microsoft/DeepSpeed/pull/1400

All is good.

Comment (thomasw21) [reply tests/test_training.py:178]: Woops sorry wrong link. Are you okay merging as is? (we can uncomment the tests as soon as the DS PR is merged + update requirements). It should not break anything (the `test_gpt` times out, thus unrelated to this PR, I guess I'll create an issue for that)

**Review by stas00 — APPROVED** [review]
Looks good - thank you for the additional commentary, Thomas

File: tests/test_training.py (hunk 5 above)
Comment (stas00) [reply tests/test_training.py:178]: Any reason not to wait till the PR on DS side is merged?
